README.md
# mod-app-manager

Copyright (C) 2022 Index Data Aps.

This software is distributed under the terms of the Apache License, Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

<!-- md2toc -l 2 README.md -->
* [Overview](#overview)
* [Invocation](#invocation)
* [Environment](#environment)
    * [`OKAPI_URL`](#okapi_url)
    * [`LOGGING_CATEGORIES`/`LOGCAT`](#logging_categorieslogcat)
* [Configuration](#configuration)
* [Descriptors](#descriptors)
* [Docker build and run](#docker-build-and-run)
* [To run under an Okapi-mediated FOLIO in a Vagrant box](#to-run-under-an-okapi-mediated-folio-in-a-vagrant-box)
* [See also](#see-also)


## Overview

This module supplies a simple WSAPI on which we can build a prototype Application Manager for FOLIO. Most importantly, it provides lists of candidate apps from a configured set of GitHub repositories such as https://github.com/MikeTaylor/example-folio-app-source. Behind the scenes, it maintains the list of GitHub repositiories that are used as sources of apps. By configuring this set of repositories, system administrators can tailor which sets of apps are made available to their users. It can be thought of as analogous to the APT source list in Debian-based Linux distributions (`/etc/apt/sources.list`).


## Invocation

`mod-app-manager` is invoked using `babel-node` to interpret ES6 features in JavaScript. The only command-line argument is the name of a configuration file: see below. Use something like:

	babel-node --presets=env -- mod-app-manager.js etc/example-config.json

The following command-line options are supported:

* `-p NUMBER` or `--port=NUMBER` -- Listen on the specified port instead of the default of 3002.
* `-V` or `--version` -- Display the version number and exit
* `-h` or `--help` -- Display a full list of options


## Environment

### `OKAPI_URL`

`mod-app-manager` stores this dynamic configuration of which GitHub sources to use in `mod-configuration`. Calls made to Okapi for this purpose are sent to Okapi that made the request, if any (using the `X-Okapi-URL` header of the incoming request), or the default URL specified in the static configuration otherwise (i.e. when used in standalone mode, not mediated by Okapi). However, if the `OKAPI_URL` environment variable is set, then this is used instead, overriding both other sources of this information.

(Why is this useful? Consider a development setup when you are side-loading `mod-app-manager` by running it in a host system, with an SSH tunnel through to a FOLIO server running in a guest machine. When requests come from that server's Okapi, the service URL as known to that Okapi is one that's private to the guest OS, such as `http://10.0.2.15:9130`, so `mod-app-manager` in the host machine can't reach it. The `OKAPI_URL` provides a way to specify an address that works from the host machine, such as `http://localhost:9130` which is typically tunnelled into the guest VM.)

### `LOGGING_CATEGORIES`/`LOGCAT`

`mod-app-manager` uses [`categorical-logger`](https://github.com/openlibraryenvironment/categorical-logger) for logging. It logs messages in the categories specified by a comma-separated list in the environment variable `LOGGING_CATEGORIES`, or if that is not defined `LOGCAT`.

The following logging categories are used:

* `config` -- logs the compiled configuration when it's parsed during startup.
* `env` -- logs relevant environment variables if they are set.
* `listen` -- logs the listen port when the server becomes available to receive connections.
* `request` -- logs each incoming request's method and path, and the response code and message
* `error` -- logs errors that are sent back in HTTP 500 responses
* `tenant` -- logs `_tenant` interface invocations when the module is enabled and disabled
* `sample` -- logs the loading process for sample records, if requested when the module is enabled

For example, you might set the categories as follows:

	export LOGCAT=config,listen,request,error


## Configuration

The single configuration file named on the command line controls how the module runs. It must be well-formed JSON consisting of a single object. At present, that can contain only a single element, `folioInfo`, which specifies details about the FOLIO system which `mod-app-manager` should run against when contacted directly. It contains the following elements:

* `url` -- The URL of the Okapi service
* `tenant` -- The tenant for which to operate
* `username` -- The name of a user to log in as at the start of the session in order to obtain a token
* `password` -- The corresponding password

Note that these configuration elements are used only when `mod-app-manager` is contacted direcly, not when access to it is mediated via Okapi. In the latter case, the URL, tenant and authentication token are taken from the incoming request (with the exception that the URL can be overridden by the `OKAPI_URL` environment variable, as noted [above](#okapi_url)).

The distribution includes [an example configuration file](etc/example-config.json).


## Descriptors

The descriptors needed for use in FOLIO (module descriptor, deployment descriptor, etc.) are generated from templates in [the `descriptors` directory](descriptors), and the expanded versions are placed in [the `target` directory](target) as with Java-based FOLIO modules.

To generate or regenerate the descriptors, use `yarn generate`.


## Docker build and run

In summary:
```sh
$ docker build -t mod-app-manager .
Sending build context to Docker daemon  67.89MB
Step 1/11 : FROM node:16-alpine AS base
[...]
Successfully built ad5ee83ded0e
Successfully tagged mod-app-manager:latest
$ docker run -dp 3002:3002 --name MAS mod-app-manager
affad42ca413579d56f9115662496eb82591995c05540e7a36dc87f09b8b987b
e$ curl -w '\n' localhost:3002/admin/health
Behold! I live!!
$ docker ps
CONTAINER ID   IMAGE           COMMAND                  CREATED          STATUS          PORTS                                       NAMES
affad42ca413   mod-app-manager "docker-entrypoint.s…"   34 seconds ago   Up 33 seconds   0.0.0.0:3002->3002/tcp, :::3002->3002/tcp   MAS
$ docker stop MAS
MAS
```


## To run under an Okapi-mediated FOLIO in a Vagrant box

Start `mod-app-manager` in the host box where you are developing:
```
shell1$ yarn start
yarn run v1.22.18
$ env LOGCAT=listen,request,error babel-node --presets=env,stage-2 -- mod-app-manager.js etc/example-config.json
(listen) mod-app-manager listening on port 3002
[...]
```

In a second shell, set up an ssh tunnel, so that the virtual machine can see the running `mod-app-manager` in the host:
```
shell2$ cd vagrant
shell2$ vagrant ssh -- -R 3002:localhost:3002
vagrant@vagrant:~$ 
[...]
```

And in a third shell, tell Okapi (via the existing tunnel) about the newly available module, and how to access it via the tunnel, and to associate it with the tenant Diku:
```
curl -w '\n' -d @target/ModuleDescriptor.json http://localhost:9130/_/proxy/modules
curl -w '\n' -d @target/Discovery.json http://localhost:9130/_/discovery/modules
curl -w '\n' -d @target/Activate.json http://localhost:9130/_/proxy/tenants/diku/modules
```

## See also

* [The MAFIA project](https://github.com/MikeTaylor/mafia), of which this is a part.
* [The FAM (FOLIO Application Metadata) file format](https://github.com/MikeTaylor/mafia/blob/master/doc/folio-app-metadata.md).
* [The register of issues](https://github.com/MikeTaylor/mod-app-manager/issues).



CHANGELOG.md
# Change history for mod-app-manager

## [1.3.0](https://github.com/MikeTaylor/mod-app-manager/tree/v1.3.0) (IN PROGRESS)

* Update the JSON Schemas and examples for FAM files to match the current [file-format definition](https://github.com/MikeTaylor/mafia/blob/master/doc/folio-app-metadata.md). Fixes #9.
* Support `_tenant` interface for `loadReference` and `loadSample` operations. Fixes #3.
* Use `id` from mod-configuration as the app-source ID. Fixes #11.

## [1.2.0](https://github.com/MikeTaylor/mod-app-manager/tree/v1.2.0) (2022-08-10)

* Drive `/app-manager/apps` from mod-configuration instead of hardwired config
* Write Okapi proxying entries for config CRUD WSAPI
* Add permissions for config CRUD, including permissions for mod-configuration
* Add instructions for runing under Okapi.
* Rename from `mod-app-store` to `mod-app-manager`
* When requests provide Okapi details (url, tenant, token) use these instead of configured
* Add `env` logging category
* Upate the Configuration section of the documentation
* Write RAML and JSON Schemas/examples for the supported web-services
* Make ID in POST response be of content-type `text/plain` with HTTP 201
* Make empty DELETE and PUT responses use HTTP status 204

## [1.1.0](https://github.com/MikeTaylor/mod-app-store/tree/v1.1.0) (2022-07-18)

* CRUD for configuration.

## [1.0.0](https://github.com/MikeTaylor/mod-app-store/tree/v1.0.0) (IN PROGRESS)

* First formal release.



src/service.js
import express from 'express';
import serveIndex from 'serve-index';
import bodyParser from 'body-parser';
import getAppsFromGitHub from './github';
import CrudToConfig from './crudToConfig';
import loadSampleRecords from './loadSampleRecords';


async function serveModAddStore(logger, port, config) {
  const returnOrReport = async (res, closure, successStatus, doNotEncode) => {
    try {
      const value = await closure();
      if (successStatus) res.status(successStatus);
      if (doNotEncode) res.contentType('text/plain');
      res.send(doNotEncode ? value : JSON.stringify(value));
    } catch (e) {
      logger.log('error', e);
      res.status(500);
      res.send(e.toString());
    }
  };


  const c2c = new CrudToConfig(config.folioInfo, 'APP_MANAGER', undefined, 'source-');
  const app = express();
  app.use(bodyParser.text({ type: () => true }));

  app.use((req, res, next) => {
    const url = req.url; // This sometimes changes by the time res.once fires
    res.once('finish', () => {
      logger.log('request', req.method, url, '-->', res.statusCode, res.statusMessage);
    });

    next();
  });

  app.get('/', (req, res) => {
    res.send(`
  This is mod-app-manager. Try:
  <ul>
    <li><a href="/admin/health">Health check</a></li>
    <li><a href="/app-manager/config/sources">Configured sources</a></li>
    <li><a href="/app-manager/apps">Apps</a></li>
    <li><a href="/htdocs/">Static area</a></li>
    <li><a href="/target/">Generated descriptors</a></li>
  </ul>
  `);
  });

  app.post('/_/tenant', async (req, res) => {
    const record = JSON.parse(req.body);
    logger.log('tenant', record);
    if (record.module_to &&
        (record.parameters || []).filter(r => r.key === 'loadSample' && r.value === 'true').length) {
      await loadSampleRecords(logger, req, c2c);
    }
    res.status(204);
    res.send();
  });

  app.use('/_/tenant/:id', (req, res) => {
    const record = req.body;
    logger.log('tenant', `ID ${req.params.id}`, record);
    res.status(204);
    res.send();
  });

  app.get('/admin/health', (req, res) => {
    res.send('Behold! I live!!');
  });

  app.get('/app-manager/apps', async (req, res) => {
    const sourceConfig = await c2c.list(req);
    returnOrReport(res, () => getAppsFromGitHub(sourceConfig));
  });

  app.get('/app-manager/config/sources', async (req, res) => {
    returnOrReport(res, () => c2c.list(req));
  });

  app.post('/app-manager/config/sources', async (req, res) => {
    const record = req.body;
    returnOrReport(res, () => c2c.add(req, record), 201, true);
  });

  app.put('/app-manager/config/sources/:id', async (req, res) => {
    const id = req.params.id;
    const record = req.body;
    returnOrReport(res, () => c2c.update(req, id, record), 204);
  });

  app.delete('/app-manager/config/sources/:id', async (req, res) => {
    const id = req.params.id;
    returnOrReport(res, () => c2c.delete(req, id), 204);
  });

  app.use('/htdocs', express.static('etc/htdocs'), serveIndex('etc/htdocs', { view: 'details' }));

  // Allow module descriptors to be accessed via HTTP, just because we can
  app.use('/target', express.static('target'), serveIndex('target', { view: 'details' }));

  app.listen(port, () => {
    logger.log('listen', `mod-app-manager listening on port ${port}`);
  });
}


export default serveModAddStore;
